### PR TITLE
test(SideNav): #17491 increased unit test coverage

### PR DESCRIPTION
**Pull request by ashna000**
**Closes** https://github.com/carbon-design-system/carbon/issues/17491

This PR adds unit tests for the SideNav component to ensure its proper rendering and functionality. These tests ensure that this component renders with the appropriate class names, accommodates custom class names, and correctly applies additional props.

**Changelog**
**Changed**

- Added Unit test file for SideNav.

**Testing / Reviewing**

- Run the test suite to ensure that all new tests pass.
- Review the tests for clarity and coverage.

**Additional Notes**
Please let me know if any further modifications are needed or if you have any questions regarding the changes.

### CODE REVIEW

**Comment by ashna000** [comment]
@kennylam , @guidari Please review this PR

**Review by guidari — APPROVED** [review]
LGTM! 🚀
I see that there are couple line that weren't tested, but I guess those lines are not possible to be tested, right?
https://app.codecov.io/gh/carbon-design-system/carbon/pull/17869/blob/packages/react/src/components/UIShell/SideNav.tsx

**Comment by ashna000** [comment]
@kennylam Please review this PR

**Comment by ashna000** [comment]
@alisonjoseph Please review this PR

**Review by guidari — CHANGES_REQUESTED** [review]
Hey @ashna000 
Looks like the Files changetestd is missing the test file for some reason. Maybe got lost in a commit.

**Comment by ashna000** [comment]
> Hey @ashna000 Looks like the Files changetestd is missing the test file for some reason. Maybe got lost in a commit.

@guidari @kennylam  It seems while updating the branch the test changes got missed. I have updated the tests. Please review.

**Comment by guidari** [comment]
Hey @ashna000 Do you think you could test the `onMouseLeave` as well?

Example in carbon: https://github.com/carbon-design-system/carbon/blob/1232b5ecad9ab650d773ff028c03ffbc9654ecd3/packages/react/src/components/Menu/Menu-test.js#L151-L156
Example in Stack Overflow: https://stackoverflow.com/questions/75952818/how-to-write-jest-test-case-for-moveover-and-mouseleave-for-functional-component#:~:text=simulate(%27-,mouseleave,-%27)%3B%0A%20%20%20%20expect

Screenshot 2024-11-08 at 16 19 15

File: packages/react/src/components/UIShell/__tests__/SideNav-test.js [hunk 1]
```diff
@@ -64,4 +69,225 @@ describe('SideNav', () => {
...
+    );
+    const sideNav = screen.getByRole('navigation');
+    fireEvent.focus(sideNav);
+    fireEvent.mouseEnter(sideNav);
+    expect(onToggle).not.toHaveBeenCalled();
+  });
+
+  it('should handle keyboard events like Escape', () => {
+    const onToggle = jest.fn();
+    render(<SideNav aria-label="test" expanded onToggle={onToggle} />);
```
Comment (ariellalgilmore) [inline packages/react/src/components/UIShell/__tests__/SideNav-test.js]:
```suggestion
    render(
      <SideNav
        aria-label="test"
        expanded
        onToggle={onToggle}
        href="#main-content"
      />
    );
```

File: packages/react/src/components/UIShell/__tests__/SideNav-test.js [hunk 2]
```diff
@@ -64,4 +69,225 @@ describe('SideNav', () => {
...
+        <TestChild />
+      </SideNav>
+    );
+    const sideNav = getByRole('navigation');
+    expect(sideNav).toHaveClass('cds--side-nav__navigation');
+    fireEvent.mouseLeave(sideNav);
+    expect(sideNav).toHaveClass('cds--side-nav__navigation');
+  });
+
+  it('should not call handleToggle if isRail is false', () => {
```
Comment (ariellalgilmore) [inline packages/react/src/components/UIShell/__tests__/SideNav-test.js:286]:
```suggestion
  it('should handleToggle if isRail is true', () => {
    const onToggleMock = jest.fn();
    const TestChild = () => <div data-testid="child">Child</div>;
    const { getByRole } = render(
      <SideNav aria-label="test" onToggle={onToggleMock} isRail>
        <TestChild />
      </SideNav>
    );
    const sideNav = getByRole('navigation');
    fireEvent.focus(sideNav);
    expect(onToggleMock).toHaveBeenCalled();
    fireEvent.mouseLeave(sideNav);
    expect(onToggleMock).toHaveBeenCalledWith(false, false);
    fireEvent.mouseEnter(sideNav);
    expect(onToggleMock).toHaveBeenCalledWith(true, true);
  });
  
  it('should not call handleToggle if isRail is false', () => {
```

File: packages/react/src/components/UIShell/__tests__/SideNav-test.js [hunk 3]
```diff
@@ -64,4 +69,225 @@ describe('SideNav', () => {
...
+      <SideNav aria-label="test" onToggle={onToggleMock} isRail>
+        <TestChild />
+      </SideNav>
+    );
+    const sideNav = getByRole('navigation');
+    expect(sideNav).not.toHaveClass('cds--side-nav--expanded');
+    fireEvent.click(sideNav);
+    expect(onToggleMock).toHaveBeenCalledWith(true, true);
+    expect(sideNav).toHaveClass('cds--side-nav--expanded');
+  });
```
Comment (ariellalgilmore) [inline packages/react/src/components/UIShell/__tests__/SideNav-test.js:316]:
```suggestion
  });
  
  it('should focus SideNav after tabbing from headerMenuButton', () => {
    render(
      <>
        <button className="cds--header__menu-toggle"></button>
        <SideNav
          aria-label="test"
          expanded
          isFixedNav={false}
          defaultExpanded
        />
      </>
    );
    const mockHeaderMenuButton = screen.getByRole('button');
    const sideNav = screen.getByRole('navigation');

    mockHeaderMenuButton.focus();
    fireEvent.keyDown(window, { key: 'Tab' });
    expect(document.activeElement).toBe(sideNav);
  });
```

**Review by ariellalgilmore — COMMENTED** [review]
Adding additional test cases like @guidari suggested

**Review by guidari — APPROVED** [review]

**Comment by guidari** [comment]
Thanks for adding that @ariellalgilmore  🚀

**Review by 2nikhiltom — APPROVED** [review]

**Pull request merged**